#### NumPy stands for Numerical Python. It is mainly used for multi-dimensional array and matrix(maths) data structure.

In [2]:
import numpy as np
print(f'Version of NumPy is: {np.__version__}')

Version of NumPy is: 1.23.5


#### One dimensional array

In [5]:
a1 = np.array([1,2,3])
print(f'One dimensional array: {a1}')

One dimensional array: [1 2 3]


#### Two Dimensional Array

In [6]:
a2 = np.array([[1,2,3],[4,5,6]])
print(f'Two dimensional array: {a2}')

Two dimensional array: [[1 2 3]
 [4 5 6]]


#### Three Dimensional Array

In [7]:
a3 = np.array([[[1,2,3],[4,5,6],[7,8,9]]])
print(f'Two dimensional array: {a3}')

Two dimensional array: [[[1 2 3]
  [4 5 6]
  [7 8 9]]]


Arrays acuire less space in memoery which makes Numpy more efficient than lists. Let's see it with the example.

In [41]:
list1 = [i for i in range(1,1000001)] # 1 to 1M 
list2 = [i for i in range(2,2000001,2)] #2 to 2M all even numbers

In [44]:
def multiply_list(a:list,b:list):
    new_list= []
    for i in range(len(a)-1):
        result = a[i] * b[i]
        new_list.append(result)
    print(new_list[0:100])

In [45]:
%%time
multiply_list(list1,list2)

[2, 8, 18, 32, 50, 72, 98, 128, 162, 200, 242, 288, 338, 392, 450, 512, 578, 648, 722, 800, 882, 968, 1058, 1152, 1250, 1352, 1458, 1568, 1682, 1800, 1922, 2048, 2178, 2312, 2450, 2592, 2738, 2888, 3042, 3200, 3362, 3528, 3698, 3872, 4050, 4232, 4418, 4608, 4802, 5000, 5202, 5408, 5618, 5832, 6050, 6272, 6498, 6728, 6962, 7200, 7442, 7688, 7938, 8192, 8450, 8712, 8978, 9248, 9522, 9800, 10082, 10368, 10658, 10952, 11250, 11552, 11858, 12168, 12482, 12800, 13122, 13448, 13778, 14112, 14450, 14792, 15138, 15488, 15842, 16200, 16562, 16928, 17298, 17672, 18050, 18432, 18818, 19208, 19602, 20000]
CPU times: user 121 ms, sys: 16 ms, total: 137 ms
Wall time: 136 ms


In [46]:
numpy1 = np.array(list1)
numpy2 = np.array(list2)

In [47]:
%%time
numpy1*numpy2

CPU times: user 2.68 ms, sys: 3.19 ms, total: 5.87 ms
Wall time: 4.01 ms


array([            2,             8,            18, ..., 1999992000008,
       1999996000002, 2000000000000])

As you can see, lists took 137ms for 1M x 1M multiplication, whereas Numpy took 5.87 ms for the same calculation. That's 2300% faster than lists.